File: apps/android-app/src/main/java/com/coinbase/cds/androidapp/gallery/GalleryNavigation.kt
package com.coinbase.cds.androidapp.gallery

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.coinbase.cds.theme.CdsTheme

/**
 * Shared chrome for full-screen gallery destinations: back affordance, title, optional prev/next
 * pager between component galleries, and a single scrollable content region.
 */
@Composable
internal fun GalleryScreenScaffold(
    title: String,
    subtitle: String?,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    testTag: String? = null,
    previousDestination: GalleryDestination? = null,
    nextDestination: GalleryDestination? = null,
    onNavigateToComponent: ((GalleryDestination) -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .background(CdsTheme.colors.bg)
            .then(if (testTag != null) Modifier.testTag(testTag) else Modifier),
    ) {
        GalleryScreenHeader(
            title = title,
            subtitle = subtitle,
            onBack = onBack,
        )
        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = CdsTheme.space.x3)
                .padding(bottom = CdsTheme.space.x2),
            verticalArrangement = Arrangement.spacedBy(CdsTheme.space.x2),
        ) {
            content()
        }
        if (onNavigateToComponent != null && (previousDestination != null || nextDestination != null)) {
            GalleryComponentPager(
                previousDestination = previousDestination,
                nextDestination = nextDestination,
                onNavigateToComponent = onNavigateToComponent,
            )
        }
    }
}

@Composable
private fun GalleryScreenHeader(
    title: String,
    subtitle: String?,
    onBack: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = CdsTheme.space.x3)
            .padding(top = CdsTheme.space.x3, bottom = CdsTheme.space.x2),
        verticalArrangement = Arrangement.spacedBy(CdsTheme.space.x1),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(CdsTheme.space.x0_5),
            modifier = Modifier
                .clickable(onClick = onBack)
                .padding(vertical = CdsTheme.space.x0_5)
                .testTag("gallery-nav-back"),
        ) {
            GalleryChevronLeftIcon(color = CdsTheme.colors.fgPrimary, iconSize = 16.dp)
            GalleryText(
                text = "Back",
                style = CdsTheme.typography.label1,
                color = CdsTheme.colors.fgPrimary,
            )
        }
        GalleryText(text = title, style = CdsTheme.typography.title1, color = CdsTheme.colors.fg)
        if (subtitle != null) {
            GalleryText(
                text = subtitle,
                style = CdsTheme.typography.body,
                color = CdsTheme.colors.fgMuted,
            )
        }
    }
}

@Composable
internal fun GalleryDestinationRow(
    title: String,
    subtitle: String,
    onClick: () -> Unit,
    testTag: String,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(CdsTheme.borderRadius.radius300))
            .background(CdsTheme.colors.bgSecondary)
            .clickable(onClick = onClick)
            .padding(CdsTheme.space.x2)
            .testTag(testTag),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(CdsTheme.space.x1),
    ) {
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(CdsTheme.space.x0_5),
        ) {
            GalleryText(text = title, style = CdsTheme.typography.headline, color = CdsTheme.colors.fg)
            GalleryText(text = subtitle, style = CdsTheme.typography.body, color = CdsTheme.colors.fgMuted)
        }
        GalleryChevronRightIcon(color = CdsTheme.colors.fgMuted, iconSize = 16.dp)
    }
}

@Composable
private fun GalleryComponentPager(
    previousDestination: GalleryDestination?,
    nextDestination: GalleryDestination?,
    onNavigateToComponent: (GalleryDestination) -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(CdsTheme.colors.bgSecondary)
            .padding(horizontal = CdsTheme.space.x3, vertical = CdsTheme.space.x1_5),
        horizontalArrangement = Arrangement.spacedBy(CdsTheme.space.x1),
    ) {
        GalleryPagerButton(
            label = previousDestination?.title?.let { "← $it" } ?: "← Previous",
            enabled = previousDestination != null,
            onClick = { previousDestination?.let(onNavigateToComponent) },
            modifier = Modifier
                .weight(1f)
                .testTag("gallery-nav-previous"),
        )
        GalleryPagerButton(
            label = nextDestination?.title?.let { "$it →" } ?: "Next →",
            enabled = nextDestination != null,
            onClick = { nextDestination?.let(onNavigateToComponent) },
            modifier = Modifier
                .weight(1f)
                .testTag("gallery-nav-next"),
        )
    }
}

@Composable
private fun GalleryPagerButton(
    label: String,
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val background = if (enabled) CdsTheme.colors.bgTertiary else CdsTheme.colors.bgSecondary
    val foreground = if (enabled) CdsTheme.colors.fg else CdsTheme.colors.fgMuted
    GalleryText(
        text = label,
        style = CdsTheme.typography.label1,
        color = foreground,
        modifier = modifier
            .clip(RoundedCornerShape(CdsTheme.borderRadius.radius300))
            .background(background)
            .then(
                if (enabled) {
                    Modifier.clickable(onClick = onClick)
                } else {
                    Modifier
                },
            )
            .padding(vertical = CdsTheme.space.x1_5, horizontal = CdsTheme.space.x1),
    )
}

File: apps/ios-gallery/Sources/GalleryNavigation.swift
@testable import CDSDesignSystem
import SwiftUI

/// Shared chrome for full-screen gallery destinations: back affordance, title, optional prev/next
/// pager between component galleries, and a single scrollable content region.
struct GalleryScreenScaffold<Content: View>: View {
    @Environment(\.cdsTheme) private var cds
    let title: String
    let subtitle: String?
    let onBack: () -> Void
    let testTag: String?
    let previousDestination: GalleryDestination?
    let nextDestination: GalleryDestination?
    let onNavigateToComponent: ((GalleryDestination) -> Void)?
    @ViewBuilder let content: Content

    init(
        title: String,
        subtitle: String? = nil,
        onBack: @escaping () -> Void,
        testTag: String? = nil,
        previousDestination: GalleryDestination? = nil,
        nextDestination: GalleryDestination? = nil,
        onNavigateToComponent: ((GalleryDestination) -> Void)? = nil,
        @ViewBuilder content: () -> Content
    ) {
        self.title = title
        self.subtitle = subtitle
        self.onBack = onBack
        self.testTag = testTag
        self.previousDestination = previousDestination
        self.nextDestination = nextDestination
        self.onNavigateToComponent = onNavigateToComponent
        self.content = content()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            ScrollView {
                VStack(alignment: .leading, spacing: cds.spacing.x2) {
                    content
                }
                .padding(.horizontal, cds.spacing.x3)
                .padding(.bottom, cds.spacing.x2)
            }
            if onNavigateToComponent != nil,
               previousDestination != nil || nextDestination != nil {
                componentPager
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(cds.colors.bg)
        .accessibilityIdentifier(testTag ?? "")
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: cds.spacing.x1) {
            Button(action: onBack) {
                HStack(spacing: cds.spacing.x0_5) {
                    Image(systemName: "chevron.left")
                        .font(.caption.weight(.semibold))
                    CDSDesignSystem.Text("Back", style: .label1, color: cds.colors.fgPrimary)
                }
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("gallery-nav-back")

            CDSDesignSystem.Text(title, style: .title1)
            if let subtitle {
                CDSDesignSystem.Text(subtitle, style: .body, color: cds.colors.fgMuted)
            }
        }
        .padding(.horizontal, cds.spacing.x3)
        .padding(.top, cds.spacing.x3)
        .padding(.bottom, cds.spacing.x2)
    }

    private var componentPager: some View {
        HStack(spacing: cds.spacing.x1) {
            GalleryPagerButton(
                label: previousDestination.map { "← \($0.title)" } ?? "← Previous",
                enabled: previousDestination != nil,
                action: { previousDestination.map { onNavigateToComponent?($0) } }
            )
            .accessibilityIdentifier("gallery-nav-previous")

            GalleryPagerButton(
                label: nextDestination.map { "\($0.title) →" } ?? "Next →",
                enabled: nextDestination != nil,
                action: { nextDestination.map { onNavigateToComponent?($0) } }
            )
            .accessibilityIdentifier("gallery-nav-next")
        }
        .padding(.horizontal, cds.spacing.x3)
        .padding(.vertical, cds.spacing.x1_5)
        .background(cds.colors.bgSecondary)
    }
}

struct GalleryDestinationRow: View {
    @Environment(\.cdsTheme) private var cds
    let title: String
    let subtitle: String
    let testTag: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(alignment: .center, spacing: cds.spacing.x1) {
                VStack(alignment: .leading, spacing: cds.spacing.x0_5) {
                    CDSDesignSystem.Text(title, style: .headline)
                    CDSDesignSystem.Text(subtitle, style: .body, color: cds.colors.fgMuted)
                }
                Spacer(minLength: 0)
                Image(systemName: "chevron.right")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(cds.colors.fgMuted)
            }
            .padding(cds.spacing.x2)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(cds.colors.bgSecondary)
            .clipShape(RoundedRectangle(cornerRadius: cds.radius.r300))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(testTag)
    }
}

private struct GalleryPagerButton: View {
    @Environment(\.cdsTheme) private var cds
    let label: String
    let enabled: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            CDSDesignSystem.Text(
                label,
                style: .label1,
                color: enabled ? cds.colors.fg : cds.colors.fgMuted
            )
            .frame(maxWidth: .infinity)
            .padding(.vertical, cds.spacing.x1_5)
            .padding(.horizontal, cds.spacing.x1)
            .background(enabled ? cds.colors.bgTertiary : cds.colors.bgSecondary)
            .clipShape(RoundedRectangle(cornerRadius: cds.radius.r300))
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
    }
}
